subsys/usb/host/usbh_class.c

```diff
@@ -0,0 +1,188 @@
+			c_node->state = USBH_CLASS_STATE_ERROR;
+		}
+	}
+
+	usbh_host_unlock(uhs_ctx);
+}
+
+void usbh_class_remove_all(struct usb_device *const udev)
+{
+	int ret;
+
+	k_mutex_lock(&udev->mutex, K_FOREVER);
+
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		struct usbh_class_data *const c_data = c_node->c_data;
+
+		if (c_data->udev == udev) {
+			ret = usbh_class_removed(c_data);
+			if (ret != 0) {
+				LOG_ERR("Failed to handle device removal for each class (%d)", ret);
+				c_node->state = USBH_CLASS_STATE_ERROR;
+				continue;
+			}
+
+			/* The class instance is now free to bind to a new device */
+			c_data->udev = NULL;
+			c_node->state = USBH_CLASS_STATE_IDLE;
+		}
+	}
+
+	k_mutex_unlock(&udev->mutex);
+}
+
+/*
+ * Probe an USB device function against all available classes of the system.
+ *
+ * Try to match a class from the global list of all system classes, using their filter rules
+ * and return status to tell if a class matches or not.
+ *
+ * The first matched class will stop the loop, and the status will be updated so that classes
+ * are only matched for a single USB function at a time.
+ *
+ * USB functions will only have one class matching, and calling usbh_class_probe_function()
+ * multiple times consequently has no effect.
+ */
+static void usbh_class_probe_function(struct usb_device *const udev,
+				      struct usbh_class_filter *const info, const uint8_t iface)
+{
+	int ret;
+
+	/* Assumes that udev->mutex is locked */
+
+	/* First check if any interface is already bound to this */
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		struct usbh_class_data *const c_data = c_node->c_data;
+
+		if (c_node->state == USBH_CLASS_STATE_BOUND &&
+		    c_data->udev == udev && c_data->iface == iface) {
+			LOG_DBG("Interface %u bound to '%s', skipping", iface, c_data->name);
+			return;
+		}
+	}
+
+	/* Then try to match this function against all interfaces */
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		struct usbh_class_data *const c_data = c_node->c_data;
+
+		if (c_node->state != USBH_CLASS_STATE_IDLE) {
+			LOG_DBG("Class %s already matched, skipping", c_data->name);
+			continue;
+		}
+
+		if (!usbh_class_is_matching(c_node->filters, info)) {
+			LOG_DBG("Class %s not matching interface %u", c_data->name, iface);
+			continue;
+		}
+
+		ret = usbh_class_probe(c_data, udev, iface);
+		if (ret == -ENOTSUP) {
+			LOG_DBG("Class %s not supporting this device, skipping", c_data->name);
+			continue;
+		}
+
+		LOG_INF("Class '%s' matches interface %u", c_data->name, iface);
+		c_node->state = USBH_CLASS_STATE_BOUND;
+		c_data->udev = udev;
+		c_data->iface = iface;
+		break;
+	}
+}
+
+void usbh_class_probe_device(struct usb_device *const udev)
+{
+	const void *desc_end;
+	const struct usb_desc_header *desc;
+	struct usbh_class_filter info;
+	uint8_t iface;
+	int ret;
+
+	k_mutex_lock(&udev->mutex, K_FOREVER);
+
+	desc = usbh_desc_get_cfg(udev);
+	desc_end = usbh_desc_get_cfg_end(udev);
```

Comment:
Drop desc_end and usbh_desc_get_cfg_end()

Reply:
done